### PR TITLE
feat: first-class dynamic tensor dimensions for streaming KV-cache decode

### LINKED ISSUE #890: [Feature]: First-class dynamic tensor dimensions for streaming KV-cache decode
# 🧠 D: DOCUMENT — Problem & Opportunity

**What is the problem, limitation, or opportunity? Why does this matter for SKaiNET?**

Autoregressive decode (Moonshine, FunctionGemma, any seq2seq/LLM) needs a **growing KV cache**: each step self-attention attends over one more token, so a tensor dimension changes every step. A `vmfb` compiled at a *fixed* cache size only serves that one step — true unbounded streaming decode needs one program that serves **every** decode position.

Today SKaiNET can only express a dynamic dimension through a downstream **post-emit text hack** (trace at a sentinel-prime dimension, then string-replace the emitted MLIR `x7919x` → `x?x`). That hack does not actually compile: the emitter, seeing static shapes, produces static op forms (`broadcast_in_dim -> tensor<…?…>`, scores-sized splat constants, `0:-1:1` slices), and the after-the-fact text substitution turns them into forms `iree-compile` rejects (`'stablehlo.broadcast_in_dim' … must be statically shaped`). So "dynamic decode" was never real.

Underlying this is an **overloaded `-1`**: it means "dynamic extent" in the emitter's `DYNAMIC_DIM`, but "infer this dim from volume" in reshape, and "last index" in the slice DSL — three incompatible meanings on one value.

**Summary (1–2 sentences):**
> Make dynamic tensor dimensions a first-class, correctly-compiling concept end-to-end — a distinct `Dim.DYNAMIC` sentinel, dynamic-aware shape arithmetic, shape-only tracing, and a dynamic-shape-safe StableHLO emitter — so a single compiled program serves every autoregressive decode position (unbounded streaming KV-cache decode).

---

# 🔍 A: ASSESS — Feasibility & Impact

### ✔️ Feasibility
Implemented and verified. The change keeps `Shape(IntArray)` storage (avoids a ~340-site `.dimensions` / ~220-site `.volume` refactor across backends) and layers a `Dim` vocabulary on top, gating every dynamic path behind `hasDynamic()` so static graphs are emitted byte-for-byte unchanged.

### ✔️ Expected Impact
- Unblocks **unbounded streaming KV-cache decode** from the pure SKaiNET DSL (no vendor binaries, any IREE runtime).
- Removes the fragile sentinel + text-substitution hack from the downstream export path.
- Both real decoders now self-compile to dynamic vmfbs: **FunctionGemma `with_past`** (dynamic `1x{nKV}x?x256`, 151 KB) and **Moonshine v2 `with_past`** (dynamic self *and* cross caches `1x8x?x40`, 86 MB weight-baked).

### ✔️ Risks / Constraints
- Numerical: none — dynamic typing does not change the computation. Verified: a growing-cache greedy decode of the Moonshine v2 dynamic vmfb matches onnxruntime **cos = 1.000000 and argmax at every step** (cache 1→12).
- API consistency: `Shape.volume` now *throws* on a dynamic shape (an unknown extent has no element count) instead of returning a corrupt product — a deliberate, fail-loud change.
- CHLO caveat: IREE's stablehlo input pipeline marks `chlo.broadcast_*` illegal, so the dynamic broadcast uses `stablehlo.dynamic_broadcast_in_dim` with a runtime `get_dimension_size` shape operand.

### ✔️ Dependencies
`skainet-lang-core` (Shape/Slice/VoidTensorOps), `skainet-compile-dag` (trace finalize), `skainet-compile-hlo` (emitter). No third-party libraries.

---

# 📚 R: RESEARCH — What Must Be Understood First?

### Research Tasks
- [x] Reproduce the exact `iree-compile` rejections a dynamic dim produces (splat constant, static `broadcast_in_dim` to `?`, `0:-1:1` slice, `?+1=0` concat).
- [x] Confirm CHLO implicit-broadcast is rejected by IREE's stablehlo pipeline → use `dynamic_broadcast_in_dim` + `get_dimension_size`.
- [x] Verify the `scale·scores == (q·scale)@kᵀ` identity to keep the scale constant at the static Q shape.
- [x] Establish blast radius of changing `Shape` representation (kept `IntArray`, layered `Dim`).

---

# 🛠️ C: CODE — Implementation Plan

### Development Tasks
- [x] `Dim` vocabulary: `DYNAMIC = Int.MIN_VALUE` (distinct from reshape `-1`), `isDynamic/isStatic/concat/compatible/render`.
- [x] `Shape`: `hasDynamic()/isDynamic(axis)/dynamicAxes`; `volume` throws on dynamic.
- [x] Slice DSL: `all()` = symbolic full-axis over a dynamic axis.
- [x] `VoidTensorOps`: allocation-free shape-only tracing (`ShapeOnlyTensorData`); concat via `Dim.concat`; reshape passes a dynamic target through.
- [x] `TraceToGraphBuilder`: don't probe `volume` of a dynamic input in finalize.
- [x] Emitter: scale-on-Q, `dynamic_broadcast_in_dim` softmax, identity reshape/slice/narrow elision, concat dynamic propagation; `TypeMapper.DYNAMIC_DIM` aliases `Dim.DYNAMIC`.
- [x] Unit tests (Dim arithmetic, guarded volume, dynamic concat/reshape/slice, dynamic SDPA emission).
- [x] Submit PR referencing this issue.

### Acceptance Criteria
- [x] Feature works and passes all tests (lang-core + compile-hlo + compile-dag suites green; static graphs byte-for-byte unchanged).
- [x] End-to-end: FunctionGemma + Moonshine v2 `with_past` self-compile to dynamic vmfbs from the DSL.
- [x] No regressions (dynamic paths gated behind `hasDynamic()`).
- [ ] Reviewed and approved by maintainers.

---

# 💬 Additional Notes

Numeric proof (Moonshine v2, dynamic vmfb vs onnxruntime `decoder_kv.onnx`, greedy lockstep):

```
step pos cache cos(ours,onnx) argmax_match
  0   1    1      1.000000        True
  1   2    2      1.000000        True
  …
 11  12   12      1.000000        True
```

### PR DESCRIPTION
**Pull request by michalharakal**
## Summary

Makes dynamic tensor dimensions a **first-class, correctly-compiling** concept end-to-end, so a single compiled `vmfb` serves every autoregressive decode position — unbounded streaming KV-cache decode from the pure SKaiNET DSL.

Previously "dynamic decode" existed only as a downstream sentinel-prime + post-emit text hack (`x7919x → x?x`) that **did not actually compile**: the emitter produced static op forms that `iree-compile` rejects once a dim is `?`. This PR fixes that at the source and removes the overloading of `-1` (which meant "dynamic" in the emitter, "infer" in reshape, and "last index" in the slice DSL).

Closes #890.

## What changed

**First-class `Dim` (`skainet-lang-core`)**
- `Dim.DYNAMIC = Int.MIN_VALUE` — a reserved sentinel **distinct from reshape's `-1` = infer** — plus centralized dynamic arithmetic (`concat`, `compatible`, `render`, `isDynamic`/`isStatic`).
- `Shape`: `hasDynamic()`, `isDynamic(axis)`, `dynamicAxes`; `volume` now throws on a dynamic shape instead of returning a corrupt product.
- Slice DSL: `all()` over a dynamic axis is a valid symbolic full-axis.

**Shape-only, dynamic-safe tracing (`skainet-lang-core`, `skainet-compile-dag`)**
- `VoidTensorOps` propagates shapes through a `ShapeOnlyTensorData` that allocates **no** backing buffer, so a `-1` extent threads through a whole decode trace (was `NegativeArraySizeException`). Concat routes through `Dim.concat`; reshape passes a dynamic target through.
- `TraceToGraphBuilder.extractFloatArray` no longer probes `volume` of a non-dense/dynamic input.

**Dynamic-shape-safe StableHLO emission (`skainet-compile-hlo`)**
- Scale folded into Q (`(q·s)@kᵀ ≡ scores·s`) — removes the invalid scores-sized dynamic splat constant.
- Softmax reduce/broadcast uses `stablehlo.dynamic_broadcast_in_dim` + a runtime `get_dimension_size` shape operand (CHLO is rejected by IREE's stablehlo pipeline).
- Identity reshape, full-extent slice/narrow elided; concat keeps the axis dynamic (`? ++ 1` stays `?`, never a bogus `0`).
- `TypeMapper.DYNAMIC_DIM` aliases `Dim.DYNAMIC` so tracer and emitter agree by construction.

Every dynamic path is gated behind `hasDynamic()`; **static graphs are emitted byte-for-byte unchanged.**

## Verification

- Full `skainet-lang-core` + `skainet-compile-hlo` + `skainet-compile-dag` suites pass; new tests cover Dim arithmetic, guarded `volume`, dynamic concat/reshape/slice, and dynamic SDPA emission.
- Synthetic dynamic SDPA `iree-compile`s and **runs at key length 3 and 17** from one module.
- **FunctionGemma `with_past`** (dynamic `1x{nKV}x?x256`, real weights) self-compiles from the DSL → 151 KB vmfb (previously uncompilable).
- **Moonshine v2 `with_past`** (dynamic self *and* cross caches `1x8x?x40`, weights baked) self-compiles → 86 MB vmfb.
- Growing-cache greedy decode of the Moonshine v2 dynamic vmfb vs onnxruntime `decoder_kv.onnx`, cache 1→12: **cos = 1.000000 and argmax match at every step.**

### CODE REVIEW

**Comment by MacOS** [comment]
You are doing a lot in this ``PR``, aren`t you? :D

**Comment by michalharakal** [comment]
Yes @MacOS    this is a quite a heavy feature, blocking SKaiNET on JVM

**Comment by MacOS** [comment]
Reviewing it is a nightmare :D.

**Comment by michalharakal** [comment]
Some features are still at an early stage and require changes across multiple layers. Missing capabilities continue to surface while building two serious, real-world applications with the library.

What gives me confidence is that the core foundation and architecture are consistently being validated. At this point, we are mainly iterating, refining, and fine-tuning the details.

**Review by aharakal — APPROVED** [review]
